server/app/controllers/admin/AdminProgramBlockPredicatesController.java

```diff
@@ -188,11 +254,196 @@ public Result updateVisibility(Request request, long programId, long blockDefini
                 leafExpression.toDisplayString(roQuestionService.getUpToDateQuestions())));
   }
 
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
+  public Result configureNewVisibilityPredicate(
+      Request request, long programId, long blockDefinitionId) {
+    requestChecker.throwIfProgramNotDraft(programId);
+
+    DynamicForm form = formFactory.form().bindFromRequest(request);
+
+    try {
+      ProgramDefinition programDefinition = programService.getProgramDefinition(programId);
+      BlockDefinition blockDefinition = programDefinition.getBlockDefinition(blockDefinitionId);
+
+      return ok(
+          predicatesConfigureView.renderVisibility(
+              request,
+              programDefinition,
+              blockDefinition,
+              getQuestionDefinitions(programDefinition, blockDefinition, form)));
+    } catch (ProgramNotFoundException | ProgramBlockDefinitionNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
+  public Result configureExistingVisibilityPredicate(
+      Request request, long programId, long blockDefinitionId) {
+    requestChecker.throwIfProgramNotDraft(programId);
+
+    try {
+      ProgramDefinition programDefinition = programService.getProgramDefinition(programId);
+      BlockDefinition blockDefinition = programDefinition.getBlockDefinition(blockDefinitionId);
+
+      ImmutableList<Long> questionIds =
+          blockDefinition
+              .visibilityPredicate()
+              .orElseThrow(
+                  () ->
+                      new RuntimeException(
+                          String.format(
+                              "Block %d has no visibility predicate", blockDefinition.id())))
+              .getQuestions();
+
+      ImmutableList<QuestionDefinition> questionDefinitions =
```

Comment:
similar suggestion to prefix with visibility as compared to the block, or program which are both input values.

Reply:
done